clients/pkg/promtail/utils/entries.go

```diff
@@ -0,0 +1,45 @@
+package utils
+
+import (
+	"sync"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+)
+
+// EntryHandlerFanouter implements api.EntryHandler, fanning out received entries to one or multiple channels.
+type EntryHandlerFanouter struct {
+	entries  chan api.Entry
+	handlers []api.EntryHandler
+
+	once sync.Once
+	wg   sync.WaitGroup
+}
+
+func NewEntryHandlerFanouter(handlers ...api.EntryHandler) *EntryHandlerFanouter {
+	multiplex := &EntryHandlerFanouter{
+		entries:  make(chan api.Entry),
+		handlers: handlers,
+	}
+	multiplex.wg.Add(1)
+	go func() {
+		defer multiplex.wg.Done()
+		for e := range multiplex.entries {
+			for _, handler := range multiplex.handlers {
+				handler.Chan() <- e
+			}
+		}
+	}()
+	return multiplex
```

Comment:
shouldn't we be running each channel send in a go routine and, have the `wg.Add(multiplex.Entries*multiplex.handlers)` and then inside each goroutinne `defer wg.Done()`

Reply:
That's actually not a bad idea, we'd have:
1. one main go routine which listens to the channel the `FanoutEntryHandler` exposes, FEH for shorten
2. upon receiving one entry, the fanout is done with one go routine per channel send
3. the go routine in 1) will only read the next message from the channel once all in 2) have completed

We need that wait until all have finished in 3), because if not, we could be reading yet another message without knowing the previous one has been fanned out properly.

Like it!